### PR TITLE
Initial dataset for a Realm

### PR DESCRIPTION
**Pull request by thesurix**
Fixes [#2254](https://github.com/realm/realm-java/issues/2254)

### CODE REVIEW

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 1]
```diff
@@ -316,6 +316,11 @@ private static void initializeRealm(Realm realm) {
                 columnInfoMap.put(modelClass, mediator.validateTable(modelClass, realm.sharedGroupManager.getTransaction()));
             }
             realm.schema.columnIndices = new ColumnIndices(columnInfoMap);
+
+            final Transaction transaction = realm.getConfiguration().getInitialDataTransaction();
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: I don't think this is enough to ensure that we only trigger this when the file is first created, e.g. I am pretty sure this will get triggered in the following case:

```
RealmConfiguration config = new RealmConfiguration.Builder(context).build();
Realm.getInstance(config).close();
Realm.getInstance(config).close(); // This will trigger it again since the schema has been cleared
```

Also, the file might be copied from assets instead. I think the only way of doing this is to have a method that checks for the existence of the Realm file and setting a flag. A good place would probably be just before calling the BaseRealm constructor.

Also, unit tests are needed to verify the behaviour of this.

File: realm/realm-library/src/main/java/io/realm/RealmConfiguration.java [hunk 2]
```diff
@@ -481,6 +495,17 @@ public Builder rxFactory(RxObservableFactory factory) {
             return this;
         }
 
+        /**
+         * Sets the initial data in {@link io.realm.Realm}. This transaction will be executed only for the first time
+         * when database file is created or while migrating the data when {@link Builder#deleteRealmIfMigrationNeeded()} is set.
+         *
+         * @param transaction transaction to execute.
+         */
+        public Builder initialData(Realm.Transaction transaction) {
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmConfiguration.java]: Note that a RealmConfiguration can be used to create both an `DynamicRealm` and a typed `Realm`. The issue didn't really describe this, but it means we probably need to support both 

```
public Builder initialData(Realm.Transaction transaction);
public Builder initialData(DynamicRealm.Transaction transaction);
```

It is slightly annoying though as the semantics gets a bit muddled (e.g. opening a DynamicRealm will not trigger the typed transaction and vice versa), but it might be the best of multiple evils.

What do you think @thesurix @realm/java ?

**Comment by cmelchior** [comment]
Hi @thesurix 
Thanks a lot for this. I really like the direction of this PR, but I am not sure it works correctly in all corner cases, so I commented on some missing unit tests that could impact the implementation.

File: realm/realm-library/src/main/java/io/realm/Realm.java (hunk 1 above)
Comment (beeender) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: by checking `version == UNVERSIONED` would probably do the trick

File: realm/realm-library/src/main/java/io/realm/RealmConfiguration.java (hunk 2 above)
Comment (kneth) [inline realm/realm-library/src/main/java/io/realm/RealmConfiguration.java]: My opinion is that supporting `Realm` only is fine. Most apps will use model classes and thereby `Realm`. Of course, it is important that we state it clearly in the documentation.

Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmConfiguration.java]: I tend to agree. We can err on the side of caution and just add support for `Realm.Transaction` now, then reevaluate if needed.

Comment (beeender) [inline realm/realm-library/src/main/java/io/realm/RealmConfiguration.java]: I don't really think user really need `public Builder initialData(DynamicRealm.Transaction transaction);`, by supplying that, it also coupled the typed Realm and dynamic Realm even more. We also have to consider a right place to trigger the initial dynamic Realm transaction when user only uses dynamic Realm. So i prefer:
1. Only initial transaction for typed Realm
2. It will only be triggered when first time create Realm (version == unversioned)

Comment (thesurix) [inline realm/realm-library/src/main/java/io/realm/RealmConfiguration.java]: Yes, I agree with @beeender. The idea behind this initial data transaction is to use `version == unversioned` condition (really simple solution). In `DynamicRealm` it is not easy to spot place where we can use this initial dataset transaction.

**Comment by thesurix** [comment]
Thanks for review. I will fix missing test soon. 

I know there is an idea about asset file, but as you mentioned we can create another issue and do it in this way :+1:

File: realm/realm-library/src/main/java/io/realm/Realm.java (hunk 1 above)
Comment (thesurix) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: @cmelchior Yes, you are right. It will be executed second time but without `realm.commitTransaction`. I think simple `version` check will do the job.

**Comment by kneth** [comment]
@thesurix Thanks in advance :-) Just notify us when you're ready.

**Comment by thesurix** [comment]
@kneth @cmelchior I fixed this PR and it is ready :)

**Comment by kneth** [comment]
Looks fine to me

**Comment by dalinaum** [comment]
:+1: Looks good to me.

**Comment by cmelchior** [comment]
Looks good to me :+1: I'll add the changelog entry in a separate commit. Thanks for adding this @thesurix

**Comment by dalinaum** [comment]
#2612 includes the changelog entry.
